### PR TITLE
Tests for work stealing

### LINKED ISSUE #305: Stealing sensitive tests
We'd like to have a couple of tests that are related to stealing sensitive workloads.

1. Very simple example, currently causing problems (https://github.com/dask/distributed/issues/6573)

```python
import time
import dask
import distributed

# dask.config.set({"distributed.scheduler.work-stealing": False})
client = distributed.Client(n_workers=4, threads_per_worker=1)
root = dask.delayed(lambda n: "x" * n)(dask.utils.parse_bytes("1MiB"), dask_key_name="root")
results = [dask.delayed(lambda *args: None)(root, i) for i in range(10000)]
dask.compute(results)
```

2. Upscaling situation (example from https://github.com/dask/distributed/issues/4471)

```python
import time
import distributed
import dask.array as da
import dask


client = distributed.Client(n_workers=1, threads_per_worker=1, memory_limit='1G')
print(client.dashboard_link)

# Slow task.
def func1(chunk):
    if sum(chunk.shape) != 0: # Make initialization fast
        time.sleep(5)
    return chunk

def func2(chunk):
    return chunk

data = da.zeros((30, 30, 30), chunks=5)
result = data.map_overlap(func1, depth=1, dtype=data.dtype)
result = result.map_overlap(func2, depth=1, dtype=data.dtype)
future = client.compute(result)

print('started computation')

time.sleep(11)
# print('scaling to 4 workers')
# client.cluster.scale(4)

time.sleep(5)
print('scaling to 20 workers')
client.cluster.scale(20)

_ = future.result()
```

3. An example that shows a very inhomogeneous compute. This should mimic a long tail in a computation graph where work stealing should reduce wall time, e.g. by using an [asymmetric distribution](https://numpy.org/doc/stable/reference/random/generated/numpy.random.lognormal.html) of sleep values

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #305

### CODE REVIEW

File: tests/benchmarks/test_work_stealing.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import distributed
+import numpy as np
+import pytest
+from coiled.v2 import Cluster
+from dask import delayed, utils
+from distributed import Client, wait
+from tornado.ioloop import PeriodicCallback
+
+
+@pytest.mark.xfail(reason="https://github.com/dask/distributed/issues/6573")
```
Comment (hendrikmakait) [inline tests/benchmarks/test_work_stealing.py:13]: We should also include this test in `distributed` once we have a fix for dask/distributed#6573

File: tests/benchmarks/test_work_stealing.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    wait(futs)
+
+    def count_work_stealing_events(dask_scheduler):
+        return len(dask_scheduler.events["stealing"])
+
+    count = small_client.run_on_scheduler(count_work_stealing_events)
+    assert count == 0
+
+
+@pytest.mark.xfail(
```
Comment (hendrikmakait) [inline tests/benchmarks/test_work_stealing.py:27]: XREF: #306

File: tests/benchmarks/test_work_stealing.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+from tornado.ioloop import PeriodicCallback
+
+
+@pytest.mark.xfail(reason="https://github.com/dask/distributed/issues/6573")
+def test_trivial_workload_should_not_cause_work_stealing(small_client):
+    root = delayed(lambda n: "x" * n)(utils.parse_bytes("1MiB"), dask_key_name="root")
+    results = [delayed(lambda *args: None)(root, i) for i in range(10000)]
+    futs = small_client.compute(results)
+    wait(futs)
+    _ = [fut.result() for fut in futs]
```
Comment (fjetter) [inline tests/benchmarks/test_work_stealing.py:19]:
```suggestion
    client.gather(futs)
```

will give the same result

File: tests/benchmarks/test_work_stealing.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    def count_work_stealing_events(dask_scheduler):
+        return len(dask_scheduler.events["stealing"])
+
+    count = small_client.run_on_scheduler(count_work_stealing_events)
+    assert count == 0
```
Comment (fjetter) [inline tests/benchmarks/test_work_stealing.py:25]: I think this is too strict. **If** we wanted to assert anything here we should rather assert that the data is distributed uniformly but we will not be able to suppress all stealing requests. This is also something we'd need to test in distributed itself.

I actually think we don't need to assert anything here. Wall time measurements will be sufficient and we will not need to xfail this test

File: tests/benchmarks/test_work_stealing.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+            time.sleep(11)
+            # print('scaling to 4 workers')
+            # client.cluster.scale(4)
+
+            time.sleep(5)
+            print("scaling to 20 workers")
+            cluster.scale(20)
+
+            _ = future.result()
```
Comment (fjetter) [inline tests/benchmarks/test_work_stealing.py:66]: You need to enter the benchmark ctx managers first. If you don't want to, use `auto_*` but in this case we do not want to measure the cluster startup so the following should be correct
```suggestion
            with benchmark_time, benchmark_task_durations, benchmark_memory:
                # Slow task.
                def func1(chunk):
                    if sum(chunk.shape) != 0:  # Make initialization fast
                        time.sleep(5)
                    return chunk

                def func2(chunk):
                    return chunk

                data = da.zeros((30, 30, 30), chunks=5)
                result = data.map_overlap(func1, depth=1, dtype=data.dtype)
                result = result.map_overlap(func2, depth=1, dtype=data.dtype)
                future = client.compute(result)

                print("started computation")

                time.sleep(11)
                # print('scaling to 4 workers')
                # client.cluster.scale(4)

                time.sleep(5)
                print("scaling to 20 workers")
                cluster.scale(20)

                _ = future.result()
```

File: tests/benchmarks/test_work_stealing.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def test_work_stealing_on_straggling_worker(test_name_uuid):
+    with Cluster(
+        name=test_name_uuid,
+        n_workers=10,
+        worker_vm_types=["t3.medium"],
+        wait_for_workers=True,
+    ) as cluster:
+        with Client(cluster) as client:
```
Comment (fjetter) [inline tests/benchmarks/test_work_stealing.py:91]: This is missing the measurements

File: tests/benchmarks/test_work_stealing.py (hunk 5 above)
Comment (fjetter) [reply tests/benchmarks/test_work_stealing.py:66]: might also be useful to define a `benchmark_all` fixture like

```python
@pytest.fixture
def benchmark_all(benchmark_memory, benchmark_time, benchmark_task_durations):
    with benchmark_memory, benchmark_time, benchmark_task_durations:
        yield
```

**Review by fjetter — CHANGES_REQUESTED** [review]

File: tests/benchmarks/test_work_stealing.py (hunk 4 above)
Comment (fjetter) [reply tests/benchmarks/test_work_stealing.py]: if it's too fast to get any meaningful signal we should remove it again. asserting on this level should be out of scope for coiled-runtime

File: tests/benchmarks/test_work_stealing.py (hunk 1 above)
Comment (fjetter) [reply tests/benchmarks/test_work_stealing.py:13]:
```suggestion
```

Please remove the xfail. This test now succeeds although it's slow.

This is "fixed by https://github.com/dask/distributed/pull/7022 already and it includes this as a test case (or a version o fit)

**Review by fjetter — APPROVED** [review]

**Comment by hendrikmakait** [comment]
The detected regressions should be unrelated given that they should be running on independent clusters and have flaked in the past.

**Pull request merged**